lib/reek/source/source_path.rb

```diff
@@ -0,0 +1,89 @@
+require 'private_attr/everywhere'
+require 'find'
+require 'pathname'
+require 'forwardable'
+
+module Reek
+  module Source
+    #
+    # Source file path in a filesystem.
+    #
+    class SourcePath
+      extend Forwardable
+      include Enumerable
+
+      def_delegators :pathname, :read, :to_s
+      def_delegator :traverse_children, :each
+
+      # Initialize with the pathname and configuration.
+      #
+      # pathname - a path as Pathname
+      def initialize(pathname, configuration: Configuration::AppConfiguration.default)
+        @configuration = configuration
+        @pathname = pathname.cleanpath
+        ensure_file
+      end
+
+      # Checks is path is relevant
+      #
+      # @return Boolean
+      def relevant?
+        ruby_file? && !ignored?
+      end
+
+      # Checks is path is ignored
+      #
+      # @return Boolean
+      def ignored?
+        path_excluded? || hidden_directory?
+      end
+
+      def ==(other)
+        other.to_s == to_s
+      end
+
+      private
+
+      private_attr_reader :configuration, :pathname
+
+      # Traverses all paths under current path, finds
+      # all relevant Ruby files and returns them as a list.
+      #
+      # @return Enumerator - if no block given
+      def traverse_children
+        return enum_for(:traverse_children) unless block_given?
+        pathname.find do |pathname|
+          if (path = SourcePath.new(pathname, configuration: configuration)).relevant?
+            yield path
```

Comment:
I think we should yield `pathname` here in order to not break the API for `Examiner`.